README.md
[![Build Status](https://github.com/UziTech/atom-jasmine3-test-runner/workflows/CI/badge.svg)](https://github.com/UziTech/atom-jasmine3-test-runner/actions)
[![Dependencies Status](https://david-dm.org/UziTech/atom-jasmine3-test-runner/status.svg)](https://david-dm.org/UziTech/atom-jasmine3-test-runner)

# Atom Jasmine 3.x Test Runner

By default, Atom runs your tests with Jasmine 1.3 (for more information on testing packages in Atom, please
[see the Atom Flight Manual](http://flight-manual.atom.io/hacking-atom/sections/writing-specs/#running-specs)).
Atom allows you to specify a custom test runner using the `atomTestRunner` field in your `package.json`, but
implementing a custom test runner is not straightforward. This module allows you to transition your specs to
Jasmine 3.x with minimal fuss.

![screenshot](./screenshots/atom-jasmine3-test-runner.gif)

## Installation

```
$ npm install [--save-dev] atom-jasmine3-test-runner
```

## Usage

### Transition from Jasmine v1.3

There is legacy support for transitioning to Jasmine 3.x from 1.3.

By default any specs with a file name matching `*-spec-v1.(js|coffee)` will be ran by the default Atom test runner after any new tests are ran.

### Default Test Runner

If you want to use all the default options, simply pass the module name as the `atomTestRunner` value in your `package.json`:

```javascript
{
  "name": "my-package",
  // ...
  "atomTestRunner": "atom-jasmine3-test-runner"
}
```

Note that your `package.json` may be cached by Atom's compile cache when running tests with Atom's GUI test runner, so
if adding or changing that field doesn't seem to work, try quitting and restarting Atom.

### Writing TypeScript Tests

⚠ NOTE: If you use a [custom runner](#Programmatic-Usage) that file (`custom-runner.js`) must be a `.js` file.

You must use [`atom-ts-transpiler`](https://www.npmjs.com/package/atom-ts-transpiler) to transpile `.ts` files to `.js` on the fly in Atom.

```bash
npm i -D atom-ts-transpiler typescript
```

In your `package.json` file you must add a `"atomTranspilers"` property.

```javascript
{
  "name": "my-package",
  // ...
  "atomTranspilers": [
    {
      "transpiler": "atom-ts-transpiler",
      "glob": "{!(node_modules)/**/,}*.ts?(x)",
      "options": {...}
    }
  ]
}
```

For more options see the [`atom-ts-transpiler` documentation](https://github.com/smhxx/atom-ts-transpiler/wiki/Setup-&-Configuration)

### Programmatic Usage

If you'd like to perform more customization of your testing environment, you can create a custom runner while still utilizing
atom-jasmine3-test-runner for most of the heavy lifting. First, set `atomTestRunner` to a _relative_ path to a file:

```javascript
{
  "name": "my-package",
  // ...
  "atomTestRunner": "./spec/custom-runner"
}
```

Then export a test runner created via the atom-jasmine3-test-runner from `./spec/custom-runner.js`:

```javascript
const { createRunner } = require('atom-jasmine3-test-runner');

// optional options to customize the runner
const extraOptions = {
  suffix: "-spec",
  legacySuffix: "-spec-v1"
};

const optionalConfigurationFunction = function() {
  // If provided, atom-jasmine3-test-runner will call this function before jasmine is started
  // so you can do whatever you'd like with the global variables.
  // (i.e. add custom matchers, require plugins, etc.)
  require("some-jasmine-plugin");

  beforeEach(function () {
    jasmine.addMatchers({
      toBeTheAnswerToTheUltimateQuestionOfLifeTheUniverseAndEverything: function (util, customEqualityTesters) {
        return {
          compare: function (actual) {
            let result = {};
            result.pass = util.equals(actual, 42, customEqualityTesters);
            const toBeOrNotToBe = (result.pass ? "not to be" : "to be"); // that is the question.
            result.message = `Expected ${actual} ${toBeOrNotToBe} the answer to the ultimate question of life, the universe, and everything.`;
            return result;
          }
        };
      }
    });
  });
}

module.exports = createRunner(extraOptions, optionalConfigurationFunction);
```

#### API

**`createRunner([options,] [callback])`**

Returns a test runner created with the given `options` and `callback`. Both parameters are optional. The returned value
can be exported from your `atomTestRunner` script for Atom to consume.

-   `options` - An object specifying customized options:

    -   `reporter` `[default: the default reporter]` - Which reporter to use on the terminal
    -   `suffix` `[default: "-spec"]` - File extension that indicates that the file contains tests
    -   `legacySuffix` `[default: "-spec-v1"]` - File extension that indicates that the file contains Jasmine v1.x tests
    -   `showColors` `[default: true]` - Colorize output on the terminal.
    -   `htmlTitle` `[default: The name of the package]` - The string to use for the window title in the HTML reporter
    -   `showEditor` `[default: false]` - Add a "Show Editor" tab to minimize the specs so you can see the editor behind it
    -   `timeReporter` `[default: false]` - Add a reporter that logs the time for each spec/suite. [TimeReporter](https://github.com/atom/atom/blob/master/spec/time-reporter.coffee)
    -   `specHelper` `[default: false]` - See [Spec Helpers](#spec-helpers)
    -   `testPaths` `[default: './spec' or './test']` - This can be an array of files to test or folders to search for files ending with the `suffix`. This does not apply to headless testing using `atom --test ./spec`
    -   `testPackages` `[default: []]` - Array of packages to include for testing. This will usually be the same packages listed in [`APM_TEST_PACKAGES`](https://github.com/atom/ci#how-do-i-install-other-atom-packages-that-my-package-build-depends-on) if using atom/ci for continuous integration  
    -   `random` `[default: false]` - Run your tests in semi-random order
    -   `seed` `[default: random]` - Sets the randomization seed if randomization is turned on
    -   `silentInstallation` `[default: false]` - Suppresses the spec helper installation messages
    -   `buildCommand` `[default: null]` - Command to build source code before loading tests. Usually `npm run build` if needed.

#### Spec Helpers
This will include Atom's [spec-helper](http://flight-manual.atom.io/hacking-atom/sections/writing-specs/#custom-matchers).

The `specHelper` option can be set to `true` to enable the spec-helper or you can specify an object enabling only the parts of the spec-helper that you want.

Example:
```javascript
const options = {
  specHelper: {
    atom: true,
    attachToDom: true,
    ci: true,
    customMatchers: true,
    jasmineFocused: true,
    jasmineJson: true,
    jasminePass: true,
    jasmineTagged: true,
    mockClock: true,
    mockLocalStorage: true,
    profile: true,
    set: true,
    unspy: true
  }
}
```

These are the possible options for `specHelper`

##### Atom
```
  atom: true
```

This will spy on `atom.menu.sendToBrowserProcess` and add default config options:

```js
atom.config.set("core.destroyEmptyPanes", false)
atom.config.set("editor.fontFamily", "Courier")
atom.config.set("editor.fontSize", 16)
atom.config.set("editor.autoIndent", false)
```

##### Attach To DOM
```
  attachToDom: true
```

This will add the function `jasmine.attachToDOM(element)` to allow you to easily attach elements to the DOM and it
takes care of removing the elements after every test so you don't need to worry about them messing with your other
tests. If you want an element to be attached to the DOM for multiple tests you can call `jasmine.attachToDOM` in a
`beforeEach` function.

##### CI
```
  ci: true
```

This will throw an error if any focused tests are left when testing in a CI environment.

This will also set `jasmine.DEFAULT_TIMEOUT_INTERVAL` to 1 minute in a CI environment.

:rotating_light: This won't do anything unless `process.env.CI` is set :rotating_light:

```yml
# .travis.yml
env:
  global:
    - CI="true"
```

```yml
# appveyor.yml
environment:
  CI: "true"
```

```yml
# circle.yml
machine:
  environment:
    CI: "true"
```

##### Custom Matchers
```
  customMatchers: true
```

Uses [jasmine2-atom-matchers](https://github.com/UziTech/jasmine2-atom-matchers)

This will add the [custom matchers](http://flight-manual.atom.io/hacking-atom/sections/writing-specs/#custom-matchers) from Atom:

-   The `toBeInstanceOf` matcher is for the `instanceof` operator
-   The `toHaveLength` matcher compares against the `.length` property
-   The `toExistOnDisk` matcher checks if the file exists in the filesystem
-   The `toHaveFocus` matcher checks if the element currently has focus
-   The `toShow` matcher tests if the element is visible in the dom

This will also include the Atom custom version of [jasmine-jquery](https://github.com/atom/atom/blob/master/vendor/jasmine-jquery.js)

##### Jasmine Focused
```
  jasmineFocused: true
```

Uses [jasmine2-focused](https://github.com/UziTech/jasmine2-focused)

This will include [jasmine-focused](https://github.com/atom/jasmine-focused#readme) (modified for Jasmine 3.x)

This includes the functions `ffdescribe`, `fffdescribe`, `ffit`, and `fffit`.

##### Jasmine JSON
```
  jasmineJson: true
```

Uses [jasmine2-json](https://github.com/UziTech/jasmine2-json)

This will include [jasmine-json](https://github.com/atom/jasmine-json#readme) (modified for Jasmine 3.x)

This includes the matcher `.toEqualJson(object)` and will give a detailed message on failure.

##### Jasmine Pass
```
  jasminePass: true
```

Uses [jasmine-pass](https://github.com/UziTech/jasmine-pass)

This will include a `pass()` function similar to Jasmine's `fail()` but opposite.

##### Jasmine Should Fail
```
  jasmineShouldFail: true
```

Uses [jasmine-should-fail](https://github.com/UziTech/jasmine-should-fail)

This will include the functions `zdescribe` and `zit` to allow you to tell jasmine that these tests should fail.

If these tests pass they will fail and if they fail they will pass but still output their messages as if they failed.

(really only useful for testing a reporter)

##### Jasmine Tagged
```
  jasmineTagged: true
```

Uses [jasmine2-tagged](https://github.com/UziTech/jasmine2-tagged)

This will include [jasmine-tagged](https://github.com/atom/jasmine-tagged#readme) (modified for Jasmine 3.x)

This includes the functions `jasmine.setIncludedTags([tags])` and `jasmine.includeSpecsWithoutTags(bool)` to allow you to filter tests easily.

##### Mock Clock
```
  mockClock: true
```

This will mock the `setTimeout` and `setInterval` functions, as well as a few others, so you can test a process that
happens on a timer with the `advanceClock` function.

When this is enabled you will need to call `jasmine.useRealClock()` if you want to use `setTimeout` or `setInterval` like usual.

This is similar to calling [jasmine.clock().install()](https://jasmine.github.io/2.9/introduction#section-Jasmine_Clock)

##### Mock Local Storage
```
  mockLocalStorage: true
```

Uses [jasmine-local-storage](https://github.com/UziTech/jasmine-local-storage)

This includes the functions `mockLocalStorage()` and `unmockLocalStorage()` to allow you to mock localStorage.

You will have to call the `mockLocalStorage()` function in-order to start mocking localStorage.

##### Profile
```
  profile: true
```

This will include the functions `measure(description, function)` and `profile(description, function)` which will write
the time the function takes to `console.log`

##### Set
```
  set: true
```

This will include the methods `.jasmineToString()` and `.isEqual(Set)` to the `Set` prototype.

##### Unspy
```
  unspy: true
```

Uses [jasmine-unspy](https://github.com/UziTech/jasmine-unspy)

This will include the function `jasmine.unspy(object, method)` to allow you to restore the original function to a spy

### Writing Tests

[Jasmine documentation](https://jasmine.github.io/)

```javascript
describe('Testing', function () {
  it('works', function () {
    expect(42).toBeTheAnswerToTheUltimateQuestionOfLifeTheUniverseAndEverything();
  });
});
```

# Credits

A huge thank you to @BinaryMuse for creating [atom-mocha-test-runner](https://github.com/BinaryMuse/atom-mocha-test-runner)
and giving me a place to start.


CHANGELOG.md
## [5.2.13](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.12...v5.2.13) (2022-05-13)


### Bug Fixes

* **deps:** update dependency glob to ^8.0.2 ([ba8f133](https://github.com/UziTech/atom-jasmine3-test-runner/commit/ba8f133aca6b6268b3c0abbd0de64500b9d770b7))

## [5.2.12](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.11...v5.2.12) (2022-04-13)


### Bug Fixes

* **deps:** update dependency glob to v8 ([cefa2f4](https://github.com/UziTech/atom-jasmine3-test-runner/commit/cefa2f49cea6904a9f558ad3822e4ba5a99c06a3))
* fix glob path separator ([392c3e9](https://github.com/UziTech/atom-jasmine3-test-runner/commit/392c3e9f269737d73642db081151a6cde1262fa8))

## [5.2.11](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.10...v5.2.11) (2022-02-15)


### Bug Fixes

* **deps:** update dependency jasmine-local-storage to ^1.1.2 ([f81b93c](https://github.com/UziTech/atom-jasmine3-test-runner/commit/f81b93cf5cba94be85ae3c2be956d80c8fa77c0c))
* **deps:** update dependency jasmine2-focused to ^1.1.2 ([ac9b251](https://github.com/UziTech/atom-jasmine3-test-runner/commit/ac9b2510b29aa2517da6560ad10ea0ee739ed648))

## [5.2.10](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.9...v5.2.10) (2022-01-04)


### Bug Fixes

* **deps:** update dependency jasmine-unspy to ^1.1.2 ([5e46655](https://github.com/UziTech/atom-jasmine3-test-runner/commit/5e466554d169f82eaa463ba2765f50d94d13e694))

## [5.2.9](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.8...v5.2.9) (2021-10-18)


### Bug Fixes

* **deps:** update dependency jasmine to ~3.10.0 ([#297](https://github.com/UziTech/atom-jasmine3-test-runner/issues/297)) ([1286ce9](https://github.com/UziTech/atom-jasmine3-test-runner/commit/1286ce98cd5808fec995f43edcaf29d2a2fbb393))

## [5.2.8](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.7...v5.2.8) (2021-08-22)


### Bug Fixes

* **deps:** update dependency jasmine to ~3.9.0 ([579672f](https://github.com/UziTech/atom-jasmine3-test-runner/commit/579672fa61d4c902ca477f20c793ba87261e3748))

## [5.2.7](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.6...v5.2.7) (2021-07-02)


### Bug Fixes

* **deps:** update dependency jasmine to ~3.8.0 ([87a4993](https://github.com/UziTech/atom-jasmine3-test-runner/commit/87a49936fbdb2b782b9506c3a7059d88431e0172))

## [5.2.6](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.5...v5.2.6) (2021-05-24)


### Bug Fixes

* **deps:** update dependency jasmine2-atom-matchers to ^1.1.9 ([0ddb7d1](https://github.com/UziTech/atom-jasmine3-test-runner/commit/0ddb7d1a9a9776a2125ffff14d6c1882ebe1b2ef))

## [5.2.5](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.4...v5.2.5) (2021-05-15)


### Bug Fixes

* **deps:** update dependency find-parent-dir to ^0.3.1 ([4947f9b](https://github.com/UziTech/atom-jasmine3-test-runner/commit/4947f9b9ea1ef659809e43b8a224842d45778018))

## [5.2.4](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.3...v5.2.4) (2021-05-06)


### Bug Fixes

* **deps:** update dependency glob to ^7.1.7 ([6a58a34](https://github.com/UziTech/atom-jasmine3-test-runner/commit/6a58a34d569acc02611843ea88b7ab10be2d28ec))

## [5.2.3](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.2...v5.2.3) (2021-04-26)


### Bug Fixes

* set env.ATOM_HOME to temp directory ([#243](https://github.com/UziTech/atom-jasmine3-test-runner/issues/243)) ([ef5e6cd](https://github.com/UziTech/atom-jasmine3-test-runner/commit/ef5e6cd96197290602a1f5294af11420e317afc2))
* use junction for windows testPackages ([43b329a](https://github.com/UziTech/atom-jasmine3-test-runner/commit/43b329ace5142c31027d67488ca5dcd4d4e88664))

## [5.2.2](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.1...v5.2.2) (2021-03-18)


### Bug Fixes

* **deps:** update dependency jasmine to ~3.7.0 ([362dabb](https://github.com/UziTech/atom-jasmine3-test-runner/commit/362dabbb6be5b55aabed849ca20185d50de22ea4))

## [5.2.1](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.2.0...v5.2.1) (2021-03-15)


### Bug Fixes

* make timeReporter optional ([9bbc775](https://github.com/UziTech/atom-jasmine3-test-runner/commit/9bbc77545dcb704326c78a160751a5111f80eb2f))

# [5.2.0](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.1.9...v5.2.0) (2021-03-15)


### Bug Fixes

* fix optionalConfigurationFunction type ([20b209b](https://github.com/UziTech/atom-jasmine3-test-runner/commit/20b209b1d7b7a047467ed4af24f686c249734f70))


### Features

* add types for createRunner ([cd4589f](https://github.com/UziTech/atom-jasmine3-test-runner/commit/cd4589f46169a23048744b0645ef18759bc5758d))
* add types for RunnerOptions ([23cd9a2](https://github.com/UziTech/atom-jasmine3-test-runner/commit/23cd9a262a4987b6cce1f4d097bc1a408bef12af))
* add types for SpecHelpers ([a7f2c99](https://github.com/UziTech/atom-jasmine3-test-runner/commit/a7f2c99d9e04dca00da41de19c06cf7c61f74d27))

## [5.1.9](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.1.8...v5.1.9) (2021-03-02)


### Bug Fixes

* **deps:** update dependency jasmine2-atom-matchers to ^1.1.8 ([3b4681a](https://github.com/UziTech/atom-jasmine3-test-runner/commit/3b4681a8b2c0ab7fecc5e82be0abdf3bc2d1a149))

## [5.1.8](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.1.7...v5.1.8) (2020-11-10)


### Bug Fixes

* **deps:** update dependency temp to ^0.9.4 ([f8959d4](https://github.com/UziTech/atom-jasmine3-test-runner/commit/f8959d4334e7ceb3e94e1b904a3e26b7281da165))

## [5.1.7](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.1.6...v5.1.7) (2020-10-30)


### Bug Fixes

* **deps:** update dependency jasmine to ~3.6.3 ([1c392a7](https://github.com/UziTech/atom-jasmine3-test-runner/commit/1c392a734ce83353d6bb569a4e63331b2ffd7b83))

## [5.1.6](https://github.com/UziTech/atom-jasmine3-test-runner/compare/v5.1.5...v5.1.6) (2020-10-26)


### Bug Fixes


types/index.d.ts
import type { TestRunner } from "atom"
//@ts-ignore
import type * as Jasmine from "jasmine"

export interface SpecHelpers {
	/**
	 * This will spy on `atom.menu.sendToBrowserProcess` and add default config options:
	 *
	 * ```js
	 * atom.config.set("core.destroyEmptyPanes", false)
	 * atom.config.set("editor.fontFamily", "Courier")
	 * atom.config.set("editor.fontSize", 16)
	 * atom.config.set("editor.autoIndent", false)
	 * ```
	 *
	 * @default `true`
	 */
	atom?: boolean

	/**
	 * This will add the function `jasmine.attachToDOM(element)` to allow you to easily attach elements to the DOM and it
	 * takes care of removing the elements after every test so you don't need to worry about them messing with your other
	 * tests. If you want an element to be attached to the DOM for multiple tests you can call `jasmine.attachToDOM` in a
	 * `beforeEach` function.
	 *
	 * @default `true`
	 */
	attachToDom?: boolean

	/**
	 * This will throw an error if any focused tests are left when testing in a CI environment.
	 *
	 * This will also set `jasmine.DEFAULT_TIMEOUT_INTERVAL` to 1 minute in a CI environment.
	 *
	 * :rotating_light: This won't do anything unless `process.env.CI` is set :rotating_light:
	 *
	 * @default `true`
	 */
	ci?: boolean

	/**
	 * Uses [jasmine2-atom-matchers](https://github.com/UziTech/jasmine2-atom-matchers)
	 *
	 * This will add the [custom
	 * matchers](http://flight-manual.atom.io/hacking-atom/sections/writing-specs/#custom-matchers) from Atom:
	 *
	 * - The `toBeInstanceOf` matcher is for the `instanceof` operator
	 * - The `toHaveLength` matcher compares against the `.length` property
	 * - The `toExistOnDisk` matcher checks if the file exists in the filesystem
	 * - The `toHaveFocus` matcher checks if the element currently has focus
	 * - The `toShow` matcher tests if the element is visible in the dom
	 *
	 * This will also include the Atom custom version of
	 * [jasmine-jquery](https://github.com/atom/atom/blob/master/vendor/jasmine-jquery.js)
	 *
	 * @default `true`
	 */
	customMatchers?: boolean

	/**
	 * Uses [jasmine2-focused](https://github.com/UziTech/jasmine2-focused)
	 *
	 * This will include [jasmine-focused](https://github.com/atom/jasmine-focused#readme) (modified for Jasmine 3.x)
	 *
	 * This includes the functions `ffdescribe`, `fffdescribe`, `ffit`, and `fffit`.
	 *
	 * @default `true`
	 */
	jasmineFocused?: boolean

	/**
	 * Uses [jasmine2-json](https://github.com/UziTech/jasmine2-json)
	 *
	 * This will include [jasmine-json](https://github.com/atom/jasmine-json#readme) (modified for Jasmine 3.x)
	 *
	 * This includes the matcher `.toEqualJson(object)` and will give a detailed message on failure.
	 *
	 * @default `true`
	 */
	jasmineJson?: boolean

	/**
	 * Uses [jasmine-pass](https://github.com/UziTech/jasmine-pass)
	 *
	 * This will include a `pass()` function similar to Jasmine's `fail()` but opposite.
	 *
	 * @default `true`
	 */
	jasminePass?: boolean

	/**
	 * Uses [jasmine-should-fail](https://github.com/UziTech/jasmine-should-fail)
	 *
	 * This will include the functions `zdescribe` and `zit` to allow you to tell jasmine that these tests should fail.
	 *
	 * If these tests pass they will fail and if they fail they will pass but still output their messages as if they failed.
	 *
	 * (really only useful for testing a reporter)
	 *
	 * @default `true`
	 */
	jasmineShouldFail?: boolean

	/**
	 * Uses [jasmine2-tagged](https://github.com/UziTech/jasmine2-tagged)
	 *
	 * This will include [jasmine-tagged](https://github.com/atom/jasmine-tagged#readme) (modified for Jasmine 3.x)
	 *
	 * This includes the functions `jasmine.setIncludedTags([tags])` and `jasmine.includeSpecsWithoutTags(bool)` to allow
	 * you to filter tests easily.
	 *
	 * @default `true`
	 */
	jasmineTagged?: boolean

	/**
	 * This will mock the `setTimeout` and `setInterval` functions, as well as a few others, so you can test a process
	 * that happens on a timer with the `advanceClock` function.
	 *
	 * When this is enabled you will need to call `jasmine.useRealClock()` if you want to use `setTimeout` or
	 * `setInterval` like usual.
	 *
	 * This is similar to calling [jasmine.clock().install()](https://jasmine.github.io/2.9/introduction#section-Jasmine_Clock)
	 *
	 * @default `true`
	 */
	mockClock?: boolean

	/**
	 * Uses [jasmine-local-storage](https://github.com/UziTech/jasmine-local-storage)
	 *
	 * This includes the functions `mockLocalStorage()` and `unmockLocalStorage()` to allow you to mock localStorage.
	 *
	 * You will have to call the `mockLocalStorage()` function in-order to start mocking localStorage.
	 *
	 * @default `true`
	 */
	mockLocalStorage?: boolean

	/**
	 * This will include the functions `measure(description, function)` and `profile(description, function)` which will
	 * write the time the function takes to `console.log`
	 *
	 * @default `true`
	 */
	profile?: boolean

	/**
	 * This will include the methods `.jasmineToString()` and `.isEqual(Set)` to the `Set` prototype.
	 *
	 * @default `true`
	 */
	set?: boolean

	/**
	 * Uses [jasmine-unspy](https://github.com/UziTech/jasmine-unspy)
	 *
	 * This will include the function `jasmine.unspy(object, method)` to allow you to restore the original function to a spy
	 *
	 * @default `true`
	 */
	unspy?: boolean
}

export interface RunnerOptions {

	/**
	 * Array of packages to include for testing. This will usually be the same packages listed in
	 * [`APM_TEST_PACKAGES`](https://github.com/atom/ci#how-do-i-install-other-atom-packages-that-my-package-build-depends-on)
	 * if using atom/ci for continuous integration
	 *
	 * @default `[]`
	 */
	testPackages?: Array<string>

	/**
	 * The `specHelper` option can be set to `true` to enable the spec-helper or you can specify an object enabling only
	 * the parts of the spec-helper that you want.
	 *
	 * @default `false` {@link [Spec Helpers](#spec-helpers)}
	 */
	specHelper?: boolean | SpecHelpers

	/**
	 * Add a reporter that logs the time for each spec/suite.
	 *
	 * @default `false` {@link [TimeReporter](https://github.com/atom/atom/blob/master/spec/time-reporter.coffee)}
	 */
	timeReporter?: boolean

	/**
	 * File extension that indicates that the file contains tests
	 *
	 * @default `"-spec"`
	 */
	suffix?: string

	/**
	 * File extension that indicates that the file contains Jasmine v1.x tests
	 *
	 * @default `"-spec-v1"`
	 */
	legacySuffix?: string

	/**
	 * Which reporter to use on the terminal
	 *
	 * @default `the Default reporter`
	 */
	reporter?: Parameters<Jasmine["addReporter"]>[0]


	/**
	 * Colorize output on the terminal.
	 *
	 * @default `true`
	 */
	showColors?: boolean

	/**
	 * The string to use for the window title in the HTML reporter
	 *
	 * @default `The Name of the package`
	 */
	htmlTitle?: string

	/**
	 * Add a "Show Editor" tab to minimize the specs so you can see the editor behind it
	 *
	 * @default `false`
	 */
	showEditor?: boolean

	/**
	 * This can be an array of files to test or folders to search for files ending with the `suffix`. This does not apply
	 * to headless testing using `atom --test ./spec`
	 *
	 * @default `'./spec' Or './test'`
	 */
	testPaths?: Array<string>

	/**
	 * Run your tests in semi-random order
	 *
	 * @default `false`
	 */
	random?: boolean

	/**
	 * Sets the randomization seed if randomization is turned on
	 *
	 * @default `Math.random()`
	 */
	seed?: number

	/**
	 * Suppresses the spec helper installation messages
	 *
	 * @default `false`
	 */
	silentInstallation?: boolean
}

/**
 * Returns a Jasmine3 test runner. The returned value can be exported from your `atomTestRunner` script for Atom to
 * consume. Both parameters are optional.
 *
 * @param options An object specifying customized options
 * @param optionalConfigurationFunction If provided, atom-jasmine3-test-runner will call this function before jasmine is
 *   started so you can do whatever you'd like with the global variables. (i.e. add custom matchers, require plugins, etc.)
 * @returns Returns a test runner created with the given `options` and `callback`.
 */
export function createRunner(options?: RunnerOptions, optionalConfigurationFunction?: () => void): TestRunner
